src/Skeleton.hs

```diff
@@ -9,132 +17,274 @@ module Skeleton
        , runSkeleton
        ) where
 
-import Control.Concurrent.Chan ( Chan, readChan )
+import Control.Concurrent.Chan ( Chan, readChan, newChan )
 import Control.Concurrent ( forkIO, threadDelay, yield )
-import Control.Monad ( forever, void )
-import Data.IORef ( IORef, modifyIORef, newIORef, readIORef )
+import Control.Monad ( forever, void, forM, forM_ )
+import Control.Monad.IO.Class ( liftIO )
+import Data.Dependent.Sum ( DSum (..) )
+import Data.Functor.Identity ( Identity (..) )
+import Data.IORef ( readIORef )
+import Data.Maybe ( catMaybes )
 import "gtk3" Graphics.UI.Gtk ( AttrOp( (:=) ) )
 import qualified "gtk3" Graphics.UI.Gtk as Gtk
 import System.Glib.Signals ( on )
 
+-- General Reflex primitives
+import Reflex.Class
+
+-- The widget-side interface for calling performEvent
+import Reflex.PerformEvent.Class
+
+-- The host-side interface for actually running actions gathered via
+-- `performEvent`
+import Reflex.PerformEvent.Base
+
+-- The widget-side interface for creating triggerable events
+import Reflex.TriggerEvent.Class
+
+-- The host-side interface for implementing triggerable events
+import Reflex.TriggerEvent.Base
+
+-- The `Spider` Reflex engine
+import Reflex.Spider ( runSpiderHost, SpiderTimeline, SpiderHost, Spider )
+
+-- Some internals from `Spider`.  I should expose these differently, so that
+-- it's not necessary to import any internal modules.
+import Reflex.Spider.Internal ( HasSpiderTimeline, SpiderHostFrame, Global )
 
 newtype GuiInputEvent = GuiInputEvent Int deriving Show
 data GuiState = GuiState (Maybe GuiInputEvent) (Maybe String) deriving Show
 
-makeStatusLabel :: IORef GuiState -> IO Gtk.Label
-makeStatusLabel stateRef = do
-  statusLabel <- Gtk.labelNew (Nothing :: Maybe String)
-  let makeStatusMessage = do
-        currentState <- readIORef stateRef
+-- | Monads that support building GTK widgets
+class Monad m => MonadGtk m where
+  -- | Lift to a context where running GTK commands is OK, as with Gtk.postGUI*
+  --TODO: Could this be more constrained than IO?
+  liftGtk :: IO a -> m a
+
+makeStatusLabel
+  :: ( -- `t` is a Reflex "timeline" - an instance of some specific Reflex engine
+       Reflex t
+       -- `m` is capable of reading temporal data in the timeline `t`; this
+       -- means that an action in `m` exists at a specific moment in time
+     , MonadSample t m
+       -- `m` can manipulate GTK objects
+     , MonadGtk m
+       -- `m` can request that actions of type `Performable m` be performed
+       -- based on `Event`s in timeline `t`.  Note that all
+       -- simultaneously-requested actions are gathered in a predictable order
+       -- and executed simultaneously; the propagation order within the Reflex
+       -- engine has no effect on the outcome
+     , PerformEvent t m
+       -- Actions requested using `PerformEvent m` can manipulate GTK objects
+     , MonadGtk (Performable m)
+     ) => Dynamic t GuiState -> m Gtk.Label
```

Comment:
The reason this needs to be a `Dynamic` instead of just an `Event` is that we need to have a way of getting the initial value - using `sample`, below.  `Event`s only have any value while they're firing, so if you miss it, you're out of luck.  `Behaviors`, on the other hand, can only be sampled - there's no way to be notified when they change.  A `Dynamic` is semantically equivalent to a pair of `(Behavior t a, Event t a)`, together with a law that the `Behavior` changes when and only when the `Event` fires.  In this case, that's exactly what we need, because that gives us enough information to ensure the GUI is up to date.

For GUIs where information is "pushed" into the GUI system, such as DOM and GTK, you'll see a lot of `Dynamic`s around, because the UI elements need to be proactive about responding to changes in the data.  For a GUI where things are periodically rendered regardless of whether they've changed, e.g. an OpenGL render loop, I'd expect to see a lot more `Behavior`s instead, since they'll be sampled at 60 FPS (or whatever the system calls for), and tracking the specific changes isn't important.  Using `Event`s only would only work in a system where the initial values are never important - perhaps something like a differential signaling protocol, where two systems are known to start in the same state and then only care to hear about changes from the other side.  I don't think this is likely to crop up in GUIs much, though.